librad/src/meta/project.rs

```diff
@@ -87,6 +154,153 @@ impl Project {
+        cleaned.signatures.clear();
+        let mut buffer: Vec<u8> = vec![];
+        let mut ser =
+            serde_json::Serializer::with_formatter(&mut buffer, CanonicalFormatter::new());
+        cleaned
+            .serialize(&mut ser)
+            .map_err(Error::SerializationFailed)?;
+        Ok(buffer)
+    }
+
+    pub fn canonical_text_contents(&self) -> Result<String, Error> {
+        String::from_utf8(self.canonical_data()?).map_err(|_| Error::InvalidUtf8)
+    }
+
+    pub fn sign(&self, key: &Key) -> Result<Signature, Error> {
+        Ok(key.sign(&self.canonical_data()?))
+    }
+
+    pub fn build_signature(&self, key: &Key) -> Result<ProjectSignature, Error> {
+        let signature = self.sign(key)?;
+        Ok(ProjectSignature {
+            key: PeerId::from(key.clone()),
+            sig: signature.encode_hex_upper(),
+        })
+    }
+
+    pub fn add_signature(&mut self, key: &Key) -> Result<(), Error> {
+        let id = PeerId::from(key.clone());
+        if self.signatures.iter().any(|s| s.key == id) {
+            Err(Error::SignatureAlreadyPresent)
+        } else if !self.maintainers.iter().any(|m| m == &id) {
+            Err(Error::SignatureFromNonMaintainer)
+        } else {
+            self.signatures.push(self.build_signature(key)?);
+            Ok(())
+        }
+    }
+
+    pub fn remove_signature(&mut self, id: &PeerId) {
+        self.signatures.retain(|s| s.key != *id);
+    }
+
+    pub fn verify_signature(&self, signature: &ProjectSignature) -> Result<(), Error> {
+        signature.verify_data(&self.canonical_data()?)
+    }
+
+    pub fn verify_signatures(&self) -> Result<(), Error> {
+        let data = self.canonical_data()?;
+        for s in self.signatures.iter() {
+            s.verify_data(&data)?
+        }
+        Ok(())
+    }
+
+    pub fn check_signatures_against_maintainers(&self) -> Result<(), Error> {
+        let mut maintainers = HashSet::<&PeerId>::new();
+        for m in self.maintainers.iter() {
+            maintainers.insert(m);
+        }
+
+        for s in self.signatures.iter() {
+            let m = maintainers.take(&s.key);
+            match m {
+                Some(_) => {}
+                None => return Err(Error::SignatureFromNonMaintainer),
+            }
+        }
+
+        if !maintainers.is_empty() {
+            return Err(Error::SignatureMissing);
+        }
+
+        Ok(())
+    }
+
+    pub fn check_validity(&self) -> Result<(), Error> {
+        self.check_signatures_against_maintainers()?;
+        self.verify_signatures()?;
+        Ok(())
+    }
+
+    pub fn is_valid(&self) -> bool {
+        match self.check_validity() {
+            Err(_) => false,
+            Ok(()) => true,
+        }
+    }
+
+    pub fn check_update(&self, previous: &Project) -> Result<(), UpdateVerificationError> {
+        if self.revision() <= previous.revision() {
+            return Err(UpdateVerificationError::NonMonotonicRevision);
```

Comment:
This happens when we are checking a whole history and the error is a `HistoryVerificationError`.
Here we are explicitly checking a given revision, so the error (`UpdateVerificationError`) implicitly refers to this revision.

More generally, it would be possible to change the `check_history` API and make it generic over the way in which we represent any given revision.
I would imagine a caller that has an iterator of git commits might want to identify each metadata revision by its git commit SHA, which at this level is simply impossible.
`check_history` would operate on an `iterator<(R,Project)>`, where `R: Display` is the generic type representing the revision.
For now I just did it the simple way.

Reply:
Well I have two thoughts around this.

1. The spec specifies that the revision is of type `uint`, so is it right to be polymorphic in revision?
2. If we were polymorphic in the revision, I think we could still carry this in the `UpdateVerificationError` variant so that it can be `Display`ed by the CI. You could impose an `Ord` instance on the revision for comparison for this function, and `Display` by the "reporting" function (which reports the error).

Reply:
My point was just that when verifying a history eventually we'd like to tell the user "at git commit $SHA the project metadata (of revision $XYZ) has problem $FOO" and not "revision $XYZ of the project metadata has problem $FOO".
For the user going from "revision $XYZ" (the `uint`) to "commit $SHA" would be inconvenient, in principle we should know the problematic SHA.

Reply:
Ah I get what you mean now :) Fair enough. I'll leave how you handle this up to you :+1: